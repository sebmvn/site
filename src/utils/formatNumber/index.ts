export const formatPriceNotation = (value: string) => {
  value = value?.replace('.', '').replace(',', '').replace(/\D/g, '')

  if (isNaN(parseFloat(value))) {
    return ''
  }

  const options = { minimumFractionDigits: 2 }
  const result = new Intl.NumberFormat('pt-BR', options).format(
    parseFloat(value) / 100
  )

  return result
}

export const formatIntegerNotation = (value: string) => {
  return value?.replace('.', '').replace(',', '').replace(/\D/g, '')
}

export const toDecimal = (value: string) => {
  const formattedValue = formatIntegerNotation(value)
  return parseFloat(formattedValue!) / 100
}

export const toInteger = (value: string) => {
  const formattedValue = formatIntegerNotation(value)
  return parseInt(formattedValue!)
}
