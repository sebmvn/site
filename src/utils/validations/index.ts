import { CalendarType } from 'hooks/use-calc'
import Joi from 'joi'

const fieldsValidations = {
  initialValue: Joi.string().required().messages({
    'any.required': 'Campo obrigatório',
    'string.empty': 'Campo obrigatório',
    'string.min': 'tamanho minimo de 1 digito',
    'string.base': 'o tipo do dado precisa ser textual'
  }),
  monthlyValue: Joi.string().required().messages({
    'any.required': 'Campo obrigatório',
    'string.empty': 'Campo obrigatório',
    'string.min': 'tamanho minimo de 1 digito',
    'string.base': 'o tipo do dado precisa ser textual'
  }),

  interestRate: Joi.string().required().messages({
    'any.required': 'Campo obrigatório',
    'string.empty': 'Campo obrigatório',
    'string.min': 'tamanho minimo de 1 digito',
    'string.base': 'o tipo do dado precisa ser textual'
  }),
  period: Joi.string().required().messages({
    'any.required': 'Campo obrigatório',
    'string.empty': 'Campo obrigatório',
    'string.min': 'tamanho minimo de 1 digito',
    'string.base': 'o tipo do dado precisa ser textual'
  }),
  interestRateCalendar: Joi.string().valid('monthly', 'yearly').required(),
  periodCalendar: Joi.string().valid('monthly', 'yearly').required()
}

export type FieldErrors = {
  [key: string]: string
}

const getFieldErrors = (objError: Joi.ValidationResult) => {
  const errors: FieldErrors = {}

  if (objError.error) {
    objError.error.details.forEach((err) => {
      errors[err.path.join('.')] = err.message
    })
  }

  return errors
}

type compoundInterestCalculatorFields = {
  initialValue: string
  monthlyValue: string
  interestRate: string
  period: string
  interestRateCalendar: CalendarType
  periodCalendar: CalendarType
}
export const compoundInterestCalculatorValidate = (
  values: compoundInterestCalculatorFields
) => {
  const schema = Joi.object(fieldsValidations)

  return getFieldErrors(schema.validate(values, { abortEarly: false }))
}
