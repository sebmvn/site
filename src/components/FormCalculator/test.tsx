import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/test-utils'
import userEvent from '@testing-library/user-event'

import FormCompoundInterestCalculator from '.'

describe('<FormCompoundInterestCalculator />', () => {
  it('should render the heading', () => {
    const { container } = renderWithTheme(<FormCompoundInterestCalculator />)

    expect(screen.getByTestId(/initialValue/i)).toBeInTheDocument()
    expect(screen.getByTestId(/monthlyValue/i)).toBeInTheDocument()
    expect(screen.getByTestId(/interestRate/i)).toBeInTheDocument()
    expect(screen.getByTestId(/period/i)).toBeInTheDocument()
    expect(screen.getByTestId(/handleSubmit/i)).toBeInTheDocument()
    expect(screen.getByTestId(/handleCleanForm/i)).toBeInTheDocument()
    expect(container.parentElement).toMatchSnapshot()
  })

  it('should show an invalid value error', async () => {
    renderWithTheme(<FormCompoundInterestCalculator />)

    userEvent.click(screen.getByTestId(/handleSubmit/i))

    expect(await screen.findAllByText(/campo obrigatÃ³rio/i)).toHaveLength(4)
  })

  it('should show an invalid value error', async () => {
    renderWithTheme(<FormCompoundInterestCalculator />)

    userEvent.type(screen.getByTestId(/initialValue/i), '5887775')

    expect(screen.getByTestId(/initialValue/i).getAttribute('value')).toBe(
      '58.877,75'
    )
  })

  it('should submit a form when values is valid', async () => {
    renderWithTheme(<FormCompoundInterestCalculator />)

    userEvent.type(screen.getByTestId(/initialValue/i), '500000')
    userEvent.type(screen.getByTestId(/monthlyValue/i), '200000s')
    userEvent.type(screen.getByTestId(/interestRate/i), '100')
    userEvent.type(screen.getByTestId(/period/i), '12')

    userEvent.click(screen.getByTestId(/handleSubmit/i))

    expect(screen.getByTestId(/initialValue/i).getAttribute('value')).toBe(
      '5.000,00'
    )
    expect(screen.getByTestId(/monthlyValue/i).getAttribute('value')).toBe(
      '2.000,00'
    )
    expect(screen.getByTestId(/interestRate/i).getAttribute('value')).toBe(
      '1,00'
    )
    expect(screen.getByTestId(/period/i).getAttribute('value')).toBe('12')
  })

  it('should clear form on click in button', async () => {
    renderWithTheme(<FormCompoundInterestCalculator />)

    userEvent.type(screen.getByTestId(/initialValue/i), '500000')

    userEvent.click(screen.getByTestId(/handleCleanForm/i))

    expect(screen.getByTestId(/initialValue/i).getAttribute('value')).toBe('')
  })
})
