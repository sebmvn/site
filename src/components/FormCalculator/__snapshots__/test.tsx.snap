// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormCompoundInterestCalculator /> should render the heading 1`] = `
.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #F4F4F4;
  border-radius: 0.5rem;
  padding: 0 1.6rem;
  border-color: #F4F4F4;
}

.c6:focus-within {
  box-shadow: 0 0 0.5rem #193241;
}

.c9 {
  color: #193241;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  padding: 0.8rem 0;
  background: transparent;
  border: 0;
  outline: none;
  padding-left: 1.6rem;
  padding-: 0;
  width: 100%;
}

.c9:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 2.4rem #F4F4F4 inset;
  -webkit-filter: none;
  filter: none;
}

.c9:-webkit-autofill::first-line {
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
}

.c4 {
  font-size: 1.4rem;
  color: #193241;
  cursor: pointer;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: #193241;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
}

.c8 > svg {
  width: 2.2rem;
  height: 100%;
}

.c8 span {
  padding: 0;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: #193241;
  -webkit-order: 1;
  -ms-flex-order: 1;
  order: 1;
}

.c12 > svg {
  width: 2.2rem;
  height: 100%;
}

.c12 span {
  padding: 0;
}

.c19 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border: 0;
  cursor: pointer;
  text-transform: uppercase;
  border-radius: 0.5rem;
  padding: 0.8rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #193241;
  font-weight: 900;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
  background: #0ADC7D;
  color: #193241;
}

.c19:hover {
  background: linear-gradient( 180deg,#0ADC7D 0%,#00823c 100% );
}

.c20 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border: 0;
  cursor: pointer;
  text-transform: uppercase;
  border-radius: 0.5rem;
  padding: 0.8rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #193241;
  font-weight: 900;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
  background: transparent;
  color: #0ADC7D;
}

.c20:hover {
  background: linear-gradient( 180deg,#0ADC7D 0%,#00823c 100% );
}

.c20:hover {
  background: transparent;
}

.c0 .c3 {
  margin: 0.8rem 0;
}

.c0 .c18 {
  margin: 3.2rem auto 1.6rem;
}

.c16 {
  background: #F4F4F4;
}

.c14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #0ADC7D;
  border-radius: 0.5rem;
  padding: 0 1.6rem;
  border-color: #F4F4F4;
  cursor: pointer;
}

.c14:focus-within {
  box-shadow: 0 0 0.5rem #193241;
}

.c15 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-appearance: button;
  cursor: pointer;
  color: #193241;
  background: #0ADC7D;
  border: 0;
  outline: 0;
  box-shadow: none;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  padding: 0.8rem 0;
  padding-left: 1.6rem;
}

.c15:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 2.4rem #F4F4F4 inset;
  -webkit-filter: none;
  filter: none;
}

.c15:-webkit-autofill::first-line {
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
}

.c15:focus {
  border: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-column-gap: 2.4rem;
  column-gap: 2.4rem;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c2 .c5 {
  padding-left: 0;
}

.c2 .c7 {
  padding: 0.8rem 1.2rem;
  color: #193241;
  background: #0ADC7D;
  border-radius: 0.5rem 0 0 0.5rem;
}

.c2 .c7 > svg {
  width: 2.5rem;
  height: 100%;
}

.c2 .c13 {
  border-radius: 0 0.5rem 0.5rem 0;
}

.c10 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c10 .c5 {
  padding-left: 0;
  padding-right: 0;
}

.c10 .c7 {
  padding: 0.8rem 1.2rem;
  color: #193241;
  background: #0ADC7D;
  border-radius: 0.5rem 0 0 0.5rem;
}

.c10 .c7 > svg {
  width: 2.5rem;
  height: 100%;
}

.c10 .c13 {
  border-radius: 0 0.5rem 0.5rem 0;
}

.c17 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-column-gap: 2.4rem;
  column-gap: 2.4rem;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c17 a {
  padding: 0;
}

.c17 .c18 {
  margin: 1.6rem 0;
}

.c11 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

@media (max-width:768px) {
  .c1 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

<body>
  <div>
    <div
      class="c0"
    >
      <form>
        <div
          class="c1"
        >
          <div
            class="c2"
          >
            <div
              class="c3 "
            >
              <label
                class="c4"
                for="initialValue"
              >
                Valor Inicial
              </label>
              <div
                class="c5 c6"
              >
                <div
                  class="c7 c8"
                >
                  <span>
                    R$
                  </span>
                </div>
                <input
                  class="c9"
                  data-testid="initialValue"
                  id="initialValue"
                  maxlength="16"
                  min="0"
                  name="initialValue"
                  placeholder="0,00"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="c2"
          >
            <div
              class="c3 "
            >
              <label
                class="c4"
                for="monthlyValue"
              >
                Valor Mensal
              </label>
              <div
                class="c5 c6"
              >
                <div
                  class="c7 c8"
                >
                  <span>
                    R$
                  </span>
                </div>
                <input
                  class="c9"
                  data-testid="monthlyValue"
                  id="monthlyValue"
                  maxlength="16"
                  name="monthlyValue"
                  placeholder="0,00"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="c1"
        >
          <div
            class="c10"
          >
            <div
              class="c3 "
            >
              <label
                class="c4"
                for="interestRate"
              >
                Taxa de Juros
              </label>
              <div
                class="c5 c6"
              >
                <div
                  class="c7 c8"
                >
                  <svg
                    aria-hidden="true"
                    class="c11"
                    fill="none"
                    focusable="false"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <line
                      x1="19"
                      x2="5"
                      y1="5"
                      y2="19"
                    />
                    <circle
                      cx="6.5"
                      cy="6.5"
                      r="2.5"
                    />
                    <circle
                      cx="17.5"
                      cy="17.5"
                      r="2.5"
                    />
                  </svg>
                </div>
                <input
                  class="c9"
                  data-testid="interestRate"
                  id="interestRate"
                  maxlength="7"
                  name="interestRate"
                  placeholder="0,00"
                  type="text"
                  value=""
                />
                <div
                  class="c12"
                >
                  <div
                    class=""
                  >
                    <div
                      class="c13 c14"
                    >
                      <select
                        class="c15"
                      >
                        <option
                          class="c16"
                          value="yearly"
                        >
                          Anual
                        </option>
                        <option
                          class="c16"
                          value="monthly"
                        >
                          Mensal
                        </option>
                      </select>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="c10"
          >
            <div
              class="c3 "
            >
              <label
                class="c4"
                for="period"
              >
                Período
              </label>
              <div
                class="c5 c6"
              >
                <div
                  class="c7 c8"
                >
                  <svg
                    aria-hidden="true"
                    class="c11"
                    fill="none"
                    focusable="false"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <rect
                      height="18"
                      rx="2"
                      ry="2"
                      width="18"
                      x="3"
                      y="4"
                    />
                    <line
                      x1="16"
                      x2="16"
                      y1="2"
                      y2="6"
                    />
                    <line
                      x1="8"
                      x2="8"
                      y1="2"
                      y2="6"
                    />
                    <line
                      x1="3"
                      x2="21"
                      y1="10"
                      y2="10"
                    />
                  </svg>
                </div>
                <input
                  class="c9"
                  data-testid="period"
                  id="period"
                  name="period"
                  placeholder="0"
                  type="text"
                  value=""
                />
                <div
                  class="c12"
                >
                  <div
                    class=""
                  >
                    <div
                      class="c13 c14"
                    >
                      <select
                        class="c15"
                      >
                        <option
                          class="c16"
                          value="yearly"
                        >
                          Anual
                        </option>
                        <option
                          class="c16"
                          value="monthly"
                        >
                          Mensal
                        </option>
                      </select>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="c17"
        >
          <button
            class="c18 c19"
            data-testid="handleSubmit"
            type="submit"
          >
            <span>
              <span>
                Calcular
              </span>
            </span>
          </button>
          <a
            class="c18 c20"
            data-testid="handleCleanForm"
          >
            <span>
              <span>
                Limpar
              </span>
            </span>
          </a>
        </div>
      </form>
    </div>
  </div>
</body>
`;
